classes/auto_login.php

```diff
@@ -0,0 +1,212 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Auto-login class.
+ *
+ * @package   auth_saml2
+ * @copyright 2020 The Open University
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace auth_saml2;
+
+use auth_saml2\admin\saml2_settings;
+
+defined('MOODLE_INTERNAL') || die();
+
+/**
+ * Auto-login class.
+ */
+class auto_login {
+    /** @var bool True if already considered auto-login on this page. */
+    protected static $considered;
+
+    /**
+     * Called on every page to give an opportunity to auto-login.
+     *
+     * This is called on most pages via the after_require_login callback, and also by the
+     * before_http_headers callback, so it might be called twice. We only want to run it once.
+     */
+    public static function consider() {
+        global $SESSION, $SCRIPT;
+
+        if (self::$considered) {
+            return;
+        }
+        self::$considered = true;
+
+        // Don't do this on any auth or login scripts!
+        if (preg_match('~^(/auth/|/login/)~', $SCRIPT)) {
+            return;
+        }
+
+        // If this is not a GET request, don't try autologin.
+        if ($_SERVER['REQUEST_METHOD'] !== 'GET') {
+            return;
+        }
+
+        // If they are already logged in, we don't need autologin.
+        if (isloggedin() && !isguestuser()) {
+            return;
+        }
+
+        // Check if this plugin is enabled.
+        $enabled = \core_plugin_manager::instance()->get_enabled_plugins('auth');
```

Comment:
There is a core get_enabled_auth_plugins. What are benefits to check if it's enabled this way?

Reply:
To be honest I wasn't aware of get_enabled_auth_plugins. The code is a bit shorter with get_enabled_plugins (because you can just look for the array key instead of having to loop through them), but performance might be marginally worse as it has to do a cache lookup, and I guess it would be more obvious to use the special-purpose function, so I'll change it to get_enabled_auth_plugins.